src/data_readers/data_reader_hdf5.cpp

```diff
@@ -0,0 +1,232 @@
+        int world_rank = get_rank_in_world();
+        int color = world_rank/dc::get_number_of_io_partitions(); 
+        MPI_Comm_split(mpi_comm, color, world_rank, &m_comm);
+        m_shuffled_indices.clear();
+        m_shuffled_indices.resize(m_file_paths.size());
+        std::iota(m_shuffled_indices.begin(), m_shuffled_indices.end(), 0);
+        std::cout<<"size after load " << m_shuffled_indices.size();
+        
+        select_subset_of_data();
+    }
+    bool hdf5_reader::fetch_label(Mat& Y, int data_id, int mb_idx) {
+        return true;
+    }
+    bool hdf5_reader::fetch_datum(Mat& X, int data_id, int mb_idx) {
+        prof_region_begin("fetch_datum", prof_colors[0], false);
+        int nprocs;
+        MPI_Comm_size(MPI_COMM_WORLD, &nprocs); 
+        if ((nprocs%dc::get_number_of_io_partitions()) !=0) {
+            std::cerr<<"if other things have not been changed in the code this will not work for anything other than 4 procs a node \n";
+        }
+        int world_rank = get_rank_in_world();
+        double start_file = MPI_Wtime();
+        //TODO: does this need to be divided by the num procs
+	    auto file = m_file_paths[data_id/dc::get_number_of_io_partitions()];
+            
+        hid_t fapl_id = H5Pcreate(H5P_FILE_ACCESS);
+        H5Pset_fapl_mpio(fapl_id, m_comm, MPI_INFO_NULL); 
+        hid_t h_file = H5Fopen(file.c_str(), H5F_ACC_RDONLY, H5P_DEFAULT);
+        if (h_file < 0) {
+            throw lbann_exception(std::string{} + __FILE__ + " " + std::to_string(__LINE__) + 
+                                    " hdf5_reader::load() - can't open file : " + file);
+        }
+
+        // load in dataset
+        hid_t h_data =  H5Dopen(h_file, HDF5_KEY_DATA.c_str(), H5P_DEFAULT);
+        hid_t filespace = H5Dget_space(h_data);
+        int rank1 = H5Sget_simple_extent_ndims(filespace);
+        hsize_t dims[rank1];
+        H5Sget_simple_extent_dims(filespace, dims, NULL);
+            
+        if (h_data < 0) {
+            throw lbann_exception(std::string{} + __FILE__ + " " + std::to_string(__LINE__) +
+                                        " hdf5_reader::load() - can't find hdf5 key : " + HDF5_KEY_DATA);
+        }
+
+        //TODO: add the int 16 stuff
+	    //TODO: change the indexing 
+	    // check if mb_idx needs to be changed to not be hard coded
+	    int adj_mb_idx = mb_idx+(world_rank%dc::get_number_of_io_partitions());
+        Mat X_v = El::View(X, El::IR(0,X.Height()), El::IR(adj_mb_idx, adj_mb_idx+1));
+
+        DataType *dest = X_v.Buffer();
+        //TODO Add a reference to X mat so only one read
+	    //will this work ? ?
+        read_hdf5(h_data, filespace, world_rank, HDF5_KEY_DATA, dims, dest);
+        //close data set
+        H5Dclose(h_data);
+        if (m_has_responses) {
+            h_data = H5Dopen(h_file, HDF5_KEY_RESPONSES.c_str(), H5P_DEFAULT);
+            H5Dread(h_data, H5T_NATIVE_DOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, m_all_responses);
```

Comment:
Since the 4 response values are stored in `double`, they need to be converted to `float`. Do something like this:
```
double responses[4];
H5Dread(h_data, ..., responses);
for (int i = 0; i < 4; ++i) {
  m_all_responses[i] = static_cast<float>(responses[i]);
}```